src/rbindlist.c

```diff
@@ -514,7 +514,16 @@ SEXP rbindlist(SEXP l, SEXP usenamesArg, SEXP fillArg, SEXP idcolArg)
         if (w==-1 || !length(thisCol=VECTOR_ELT(li, w))) {  // !length for zeroCol warning above; #1871
           writeNA(target, ansloc, thisnrow);  // writeNA is integer64 aware and writes INT64_MIN
         } else {
-          if ((TYPEOF(target)==VECSXP || TYPEOF(target)==EXPRSXP) && TYPEOF(thisCol)!=TYPEOF(target)) {
+          if ((TYPEOF(target)==VECSXP) && TYPEOF(thisCol)>TYPEOF(target)) {
+            // Exotic non-atomic types need each element to be wrapped in a list, e.g. expression vectors #546
+            SEXP thisColList = PROTECT(allocVector(VECSXP, length(thisCol))); nprotect++;
+            for(int r=0; r<length(thisCol); ++r) {
+              SEXP thisElement = VECTOR_ELT(thisCol, r);
+              if (TYPEOF(thisCol) == EXPRSXP) thisElement = coerceVector(thisElement, EXPRSXP); // otherwise LANGSXP
```

Comment:
Are there other more exotic types that people use in columns? It would be good to test whether we need to always do this additional coercion step, or whether this problem is unique to EXPRSXP vectors.

Reply:
We need to protect here as well. Test for both language object created with `quote` and expressions would be nice. Regarding consistency to base R, it is best to evaluate that taking use cases in account. Language objects IMO has quite the same use case as expressions, not sure what base R behaviour on expressions is meant to archive.

Reply:
So a PROTECT for both the `coerceVector` and the `VECTOR_ELT`? Or just the `coerceVector` (this line is the same as one referred to in my review comment below).

Reply:
VECTOR_ELT is just a pointer to an object already referenced, thus protected. Just result of coerceVector.

Reply:
I'm not sure we need to handle `quote`. Constructing a quote column on its own does strange things because its length is not necessarily one, it depends on the parse tree I think. E.g. 

```
> length(quote(as.character(system.time())))
[1] 2
> length(quote(as.character(system.time()+5)))
[1] 2
> length(quote(c(1, 2, 3)))
[1] 4
> length(quote(1+1))
[1] 3
```

If you try to create a vector of `quote` it turns into a list already (conveniently the desired behaviour in data.table)

```
> c(quote(1+1), quote(as.character(system.time())))
[[1]]
1 + 1

[[2]]
as.character(system.time())
```

In the event of a nrow 1 data.table, all your other columns get recycled to the length of the quote object, which I think is probably not desirable behaviour:

```
> data.table(A=0, B=quote(1+1))
       A      B
   <num> <call>
1:     0  1 + 1
2:     0  1 + 1
3:     0  1 + 1
```

Reply:
RE: base R. I'm referring to #4040 where the person has asked for data.table NOT to turn each element into a list.

Reply:
IMO expressions should be wrapped into list as they are not atomic, as as language objects. Quote is length 1 only for symbols, otherwise it is a call, still language object, which length depends on the number of arguments.

Reply:
I believe the latest commit solves this, but will need to add tests